pkg/controller/eventtype/reconcile.go

```diff
@@ -0,0 +1,215 @@
+		return reconcile.Result{}, nil
+	}
+
+	// If the EventType exists but could not be retrieved, then we should retry.
+	if err != nil {
+		glog.Errorf("could not fetch EventType %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	// Now that we know the EventType exists, we can reconcile it. An error returned
+	// here means the reconcile did not complete and the EventType should be requeued
+	// for another attempt.
+	// A successful reconcile does not necessarily mean the EventType is in the desired
+	// state, it means no more can be done for now. In this case the EventType will
+	// not be reconciled again until the resync period or a watched resource changes.
+	if err = r.reconcile(ctx, et); err != nil {
+		glog.Errorf("error reconciling EventType: %v", err)
+	}
+
+	// Since the reconcile is a sequence of steps, earlier steps may complete
+	// successfully while later steps fail. The EventType is updated on failure to
+	// preserve any useful status or metadata changes the non-failing steps made.
+	if updateErr := r.updateEventType(ctx, et); updateErr != nil {
+		glog.Errorf("failed to update EventType: %v", updateErr)
+		// An error here means the EventType should be reconciled again, regardless of
+		// whether the reconcile was successful or not.
+		return reconcile.Result{}, updateErr
+	}
+	return reconcile.Result{}, err
+}
+
+// reconcile tries to converge the current state of the given EventType to the desired state. This
+// function should not update the EventType in the API server. This method will update 'et'. The
+// calling method is responsible for writing back to the API server.
+func (r *reconciler) reconcile(ctx context.Context, et *feedsv1alpha1.EventType) error {
+	if et.GetDeletionTimestamp() == nil {
+		r.addFinalizer(et)
+	} else {
+		err := r.handleDeletion(ctx, et)
+		if err != nil {
+			glog.Errorf("Error deleting the EventType: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *reconciler) addFinalizer(et *feedsv1alpha1.EventType) {
+	finalizers := sets.NewString(et.Finalizers...)
+	finalizers.Insert(finalizerName)
+	et.Finalizers = finalizers.List()
+}
+
+func (r *reconciler) removeFinalizer(et *feedsv1alpha1.EventType) {
+	finalizers := sets.NewString(et.Finalizers...)
+	finalizers.Delete(finalizerName)
+	et.Finalizers = finalizers.List()
+}
+
+// handleDeletion checks the finalizer conditions of an EventType marked for deletion. If the
+// conditions are met, then it removes the finalizer. Otherwise it adds a Status saying why it
+// can't.
+func (r *reconciler) handleDeletion(ctx context.Context, et *feedsv1alpha1.EventType) error {
+	feeds, err := r.findFeedsUsingEventType(ctx, et)
+	if err != nil {
+		glog.Infof("Unable to find Feeds using EventType %s: %v", et.Name, err)
+		return err
+	}
+	if len(feeds) == 0 {
+		r.removeFinalizer(et)
+	} else {
+		glog.Infof("Cannot remove finalizer from EventType %q, %v Feed(s) still use it.", et.Name, len(feeds))
```

Comment:
would be better to add a condition to the eventtype to get this message to the k8s user who issued the delete and is wondering why the object is not going away.

Reply:
That is happening in `r.updateInUseStatus()` on line 126.

```
  status:
    conditions:
    - message: 'Still in use by the Feeds: k8s-event-bar-flow-r5bf2, k8s-event-foo-flow-48nzm'
      status: "True"
      type: InUse
```

Reply:
ah, then I would just use that message to log, don't make two messages that are similar.